<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>MCGraphs · MCGraphs.jl v0.1.0</title><meta name="title" content="MCGraphs · MCGraphs.jl v0.1.0"/><meta property="og:title" content="MCGraphs · MCGraphs.jl v0.1.0"/><meta property="twitter:title" content="MCGraphs · MCGraphs.jl v0.1.0"/><meta name="description" content="Documentation for MCGraphs.jl v0.1.0."/><meta property="og:description" content="Documentation for MCGraphs.jl v0.1.0."/><meta property="twitter:description" content="Documentation for MCGraphs.jl v0.1.0."/><script data-outdated-warner src="assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="assets/documenter.js"></script><script src="search_index.js"></script><script src="siteinfo.js"></script><script src="../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="index.html">MCGraphs.jl v0.1.0</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li class="is-active"><a class="tocitem" href="index.html">MCGraphs</a><ul class="internal"><li class="toplevel"><a class="tocitem" href="#Index"><span>Index</span></a></li></ul></li><li><a class="tocitem" href="validations.html">Validations</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href="index.html">MCGraphs</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href="index.html">MCGraphs</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/tanaylab/MCGraphs.jl/blob/main{path}?plain=1#L{line}" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="MCGraphs"><a class="docs-heading-anchor" href="#MCGraphs">MCGraphs</a><a id="MCGraphs-1"></a><a class="docs-heading-anchor-permalink" href="#MCGraphs" title="Permalink"></a></h1><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MCGraphs.MCGraphs" href="#MCGraphs.MCGraphs"><code>MCGraphs.MCGraphs</code></a> — <span class="docstring-category">Module</span></header><section><div><p>Generate graphs for visualizing scRNA-seq metacells data in a <code>Daf</code> data set.</p><p>The goals of this are:</p><ul><li>Provide a specific set of graphs allowing for investigation and visualization of the metacells data.</li><li>Generate static PNG and/or SVG files (for publishing).</li><li>Generate interactive graphs in Jupyter notebook, both from Python and from R (for exploration).</li><li>Provide guided UI in Jupyter notebook for accessing and customizing the graphs, as well as generating standard dashboards of related graphs.</li><li>Be reasonably efficient when the data is large.</li></ul><p>The overall architecture of the solution is based on the following layered implementation:</p><p>Julia code:</p><ul><li><a href="validations.html#MCGraphs.Validations"><code>Validations</code></a> contains functions for validating user input which can be invoked by the UI to ensure valid inputs.</li><li><code>Renderers</code> contains functions that actually render graphs, given the graph data and configuration. Each basic graph type has its own specification for the data and for the configuration. For example, a scatter plot data contains x coordinates, y coordinates, and colors, while the configuration controls the graph size, point shape and size, axis and graph titles, etc. We render most graphs using <code>Plotly</code>, but use <code>ClusterGrammer</code> for heatmaps.</li><li><code>Extractors</code> contains functions that extract graph data from a <code>Daf</code> data set. For example, a gene-gene plot will extract the expression of two genes in all metacells as the x and y coordinates, colored according to the type of the metacells.</li><li><code>Plotters</code> contains functions that combine extraction and rendering, that is, will render a graph given a <code>Daf</code> data set and a graph configuration. Each plot can specify different defaults for the graph configuration. For example, the default titles of a gene-gene plot would be different from the default titles of a differential expression plot, even though both will use the same scatter plot renderer.</li></ul><p>Jupyter notebook code (need to be implemented in all Jupyter languages):</p><ul><li><code>Controllers</code> are Jupyter notebook UI elements for specifying which data to extract for a graph, or the configuration of a graph. For example, a data controller for a gene-gene plot will allow selecting the two genes, and a configuration controller for a scatter plot graph will allow specifying the point shape and size.</li><li><code>Widgets</code> combine multiple controllers to fully specify a graph to render. For example, a gene-gene graph widget will combine both the gene-gene plot data selection with the scatter plot graph configuration, using the specific defaults of the gene-gene graph.</li><li><code>Menus</code> allow selecting a single widget (for rendering a single graph) or several related widgets (for rendering a standard dashboard of graphs). For example, one entry in a menu could be &quot;gene-gene plot&quot; which generates a single graph while another might be &quot;metacells QC dashboard&quot; which generates multiple graphs visualizing various metacells QC measures. Menus are nested in a tree, such that invoking the root menu allows the user to interactively navigate to any of the provided plots and dashboards, without having to memorize the function names and configuration options.</li></ul><p>Generating graphs is in general a read-only operation, allowing the analyst to explore the data. However, some work flows require creating and modifying data, for example, manual metacells type annotations. This is supported by dedicated controllers that write data back into the <code>Daf</code> data set, which is typically a chain consisting of a large read-only base data set combined with a small writeable data set containing the manually entered data. This allows reusing the same base data set with multiple alternative manual annotations.</p><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>While the overall architecture of the code here generalizes well, the functionality here is intentionally restricted to what we found useful for visualizing scRNA-seq metacells data in <code>Daf</code> data sets. Likewise, the amount of customization of the graphs is intentionally limited. The intent here is to make it as easy as possible for the analyst working in Jupyter notebook to explore the data, and generate graphs for academic papers, <strong>not</strong> to create yet another end-all-be-all graphs framework (of which there are too many already).</p></div></div><p>This is intentionally separated from the [<code>Metacells.jl</code>]((https://github.com/tanaylab/Metacells.jl) package itself, which is dedicated to computing the metacells.</p></div></section></article><h1 id="Index"><a class="docs-heading-anchor" href="#Index">Index</a><a id="Index-1"></a><a class="docs-heading-anchor-permalink" href="#Index" title="Permalink"></a></h1><ul><li><a href="index.html#MCGraphs.MCGraphs"><code>MCGraphs.MCGraphs</code></a></li><li><a href="validations.html#MCGraphs.Validations"><code>MCGraphs.Validations</code></a></li><li><a href="validations.html#MCGraphs.Validations.ObjectWithValidation"><code>MCGraphs.Validations.ObjectWithValidation</code></a></li><li><a href="validations.html#MCGraphs.Validations.assert_valid_object"><code>MCGraphs.Validations.assert_valid_object</code></a></li><li><a href="validations.html#MCGraphs.Validations.validate_object"><code>MCGraphs.Validations.validate_object</code></a></li></ul></article><nav class="docs-footer"><a class="docs-footer-nextpage" href="validations.html">Validations »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.4.0 on <span class="colophon-date" title="Thursday 18 April 2024 11:37">Thursday 18 April 2024</span>. Using Julia version 1.10.0.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
